### PR TITLE
FOGL-5572 Filter rename functionality added

### PR DESCRIPTION
**Pull request by ashish-jabble**
Signed-off-by: ashish-jabble <ashish@dianomic.com>

**Packaging contents**

_Debian_

```
$ sudo dpkg -c archive/DEBIAN/x86_64/fledge-filter-rename-1.9.1-x86_64.deb 
[sudo] password for aj: 
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/python/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/python/fledge/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/python/fledge/plugins/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/python/fledge/plugins/filter/
drwxrwxr-x aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/python/fledge/plugins/filter/rename/
-rw-rw-r-- aj/aj            55 2021-06-29 15:32 ./usr/local/fledge/python/fledge/plugins/filter/rename/VERSION.filter.rename
-rw-rw-r-- aj/aj             0 2021-06-29 15:32 ./usr/local/fledge/python/fledge/plugins/filter/rename/__init__.py
-rw-rw-r-- aj/aj          5682 2021-06-29 15:32 ./usr/local/fledge/python/fledge/plugins/filter/rename/rename.py
```

_RPM_
```
$ rpm -qlp archive/x86_64/fledge-filter-rename-1.9.1-1.x86_64.rpm 
/usr/local/fledge/python/fledge/plugins/filter/rename/VERSION.filter.rename
/usr/local/fledge/python/fledge/plugins/filter/rename/__init__.py
/usr/local/fledge/python/fledge/plugins/filter/rename/rename.py

```

### CODE REVIEW

**Review by MarkRiddoch — APPROVED** [review]
Looks good to me, either add the documentation in this Jira or create another if you prefer.

File: python/fledge/plugins/filter/rename/rename.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        new_dict['asset'] = re.sub(search_pattern, replace_with, new_dict['asset'], flags=re.IGNORECASE)
+    elif operation == 'datapoint':
+        dp = re.sub(search_pattern, replace_with, str(new_dict['readings']), flags=re.IGNORECASE)
+        new_dict['readings'] = eval(dp)
+    elif operation == 'both':
+        # Both asset and datapoint case
+        both = re.sub(search_pattern, replace_with, str(reading), flags=re.IGNORECASE)
+        new_dict = eval(both)
+    else:
+        _LOGGER.warning("Unknown {} operation found.")
```
Comment (praveen-garg) [inline python/fledge/plugins/filter/rename/rename.py:178]: _LOGGER.warning("Unknown operation %s, forwarding the readings as is.", operation)

File: python/fledge/plugins/filter/rename/rename.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    # Filter is enabled: compute for each reading
+    processed_data = {}
+    for element in data:
+        processed_data = find_and_replace(handle['operation']['value'], handle['find']['value'],
+                                          handle['replaceWith']['value'], element)
+    _LOGGER.debug("processed data {}".format([processed_data]))
+    # Pass data onwards
+    filter_ingest.filter_ingest_callback(the_callback, the_ingest_ref, [processed_data])
+
+    _LOGGER.info("{} filter ingest done".format(PLUGIN_NAME))
```
Comment (praveen-garg) [inline python/fledge/plugins/filter/rename/rename.py:152]: better not called ingest (in logs at-least) what if applied on north. We should actually avoid this log statement as INFO (as of plugin) for performance reasons, can keep as debug/ better commented.

File: python/fledge/plugins/filter/rename/rename.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+import copy
+import logging
+
+from fledge.common import logger
+import filter_ingest
+
+__author__ = "Ashish Jabble"
+__copyright__ = "Copyright (c) 2021 Dianomic Systems Inc."
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
```
Comment (praveen-garg) [inline python/fledge/plugins/filter/rename/rename.py:19]: redundant?! We didn't use module level dunder names, I think.

File: python/fledge/plugins/filter/rename/rename.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    Args:
+        operation:     Possible values are asset, datapoint or both
+        find:          A regular expression to match for the given operation
+        replace_with:  A substitution string to replace the matched text with
+        reading:       A reading object
+    Returns:
+        dict:          A processed dictionary
+    """
+    new_dict = reading.copy()
+    search_pattern = r'{}'.format(find)
```
Comment (praveen-garg) [inline python/fledge/plugins/filter/rename/rename.py:165]: Let's keep that boundary (match exact name) check. And we can merge it. Other improvements can be taken up as needed via other Jira.

Comment (ashish-jabble) [reply python/fledge/plugins/filter/rename/rename.py]: Done

**Comment by ashish-jabble** [comment]
Read the docs documentation is also added in this PR

**Review by praveen-garg — APPROVED** [review]

**Pull request merged**